mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp

```diff
@@ -501,8 +502,16 @@ struct BlockwiseGemmV2RewritePattern
 
       // for(index_t k_i = 0; k_i < KPerThread; k_i += mfma_type.num_input_blks)
       // {
-      //     a[k_i] = p_a_wave[(k_i * num_input_blks + blk_id) * M + blk_td];
-      //     b[k_i] = p_b_wave[(k_i * num_input_blks + blk_id) * N + blk_td];
+      //   for (index_t mr = 0; mr < mRepeats; mr++)
+      //   {
+      //     a[mr*KPerThread + k_i] = p_a_wave[(k_i * num_input_blks + blk_id) *
+      //     M + blk_td + mr*num_input_blks];
+      //   }
+      //   for (index_t nr = 0; nr < nRepeats; nr++)
+      //   {
+      //     b[nr*KPerThread + k_i] = p_b_wave[(k_i * num_input_blks + blk_id) *
+      //     N + blk_td + nr*num_input_blks];
+      //   }
```

Comment:
I'd prefer:
 - Repeat loop stay outside, and K loop stay inside
 - Once above is done, fuse it with the non-reduction case for the outer and inner loop. We only do the code generation differently inside the inner loop, based on reduction/non-reduction.

Reply:
So something like:
```
for m=0:mRepeat:
   for n=0:nRepeat:
      for k=0:KPerThread:
```

Reply:
I think whether you want `m/n` or `k` being the inner loop depends on the layout of LDS, which means it's a function of `kpack`.

Or possibly this is another case for `getMaxVectorization()` to determine the loop order.

Reply:
Looking at non-reduction cases, I think I get what @jerryyin meant. He wants the code to look exactly like non-reduction, where they use the second version of the loop (i.e., the two separate loops). Is that right @jerryyin ? Btw, why is it a function of KPack? Still looking at the non-reduction case, I don't see much different in the logic they treat `KPack==1` and `KPack>1` cases, except for a multiplication by `KPack` itself when it's `>1`.

Reply:
I mention kPack because it affects the LDS layout, which is `[k, m, kPack]`. If a thread needs to load multiple `k` values, say `A[0, 0]` and `A[0, 1]` (logical m/k coordinates), it matters if those are `buffer[0]` and `buffer[1]` (as they would be if `kPack >= 2`) or if they are `buffer[0]` and `buffer[M]`, as they would be if `kPack = 1`.

Reply:
Well, yes, but if `KPack  > 1`  I think we always deal with vectors of KPack and those vectors are always a stride distant (M or N). Basically (IIUC) the unit of load is always a stride apart in the LDS buffer

Reply:
> for k = 0:KPerThread:
    for m = 0:mRepeat:
        load A[k,m]
    for n = 0:nRepeat:
        load B[k,n]

Please avoid this. This will make you run into the same backend bug as we have discussed yesterday. Have two loops and use lambda in the inner loop.

Reply:
@krzysz00 
> I mention kPack because it affects the LDS layout

Those current loop computation will consider about kPack but hardcoded multiplication from base offset with kPack. I think the biggest obstacle we fuse those loop together is blocked by backend bug.

> Or possibly this is another case for getMaxVectorization() to determine the loop order.

Yes, once we move kPack to be its own dimension, using max vectorization is a meaningful thing to try.